File: P2.py
## Problem2: Vertical order tree traversal https://leetcode.com/problems/binary-tree-vertical-order-traversal/
# 3 
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def verticalOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        self.d = {}
        if not root:
            return None
        def bfs_n_dfs(root, col, h):
            # print(self.d)
            if not root:
                return
            if col not in self.d:
                self.d[col] = [[h, root.val]]
            else:
                self.d[col].append([h, root.val])
            if root.left:
                bfs_n_dfs(root.left, col - 1, h + 1)
            if root.right:
                bfs_n_dfs(root.right, col + 1, h + 1)
    
        bfs_n_dfs(root, 0, 0)
        i = min(self.d.keys())
        j = max(self.d.keys())
        res = []
        for k in range(i, j+1):
            self.d[k].sort()
            d1 = [a[1] for a in self.d[k]]
            # d1.sort()
            res.append(d1)
        
        # print(self.d)
        return res

File: p2.java
// Time Complexity :O(n)
// Space Complexity :O(n)
// Did this code successfully run on Leetcode :yes
// Any problem you faced while coding this :no


// Your code here along with comments explaining your approach

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Codec {

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        if(root == null) return "";
        StringBuilder sb = new StringBuilder();
        Queue<TreeNode> q = new LinkedList<>();
        //bfs
        q.add(root);
        while(!q.isEmpty()){
            TreeNode node = q.remove();
            if(node.val == 1001){
                sb.append("#,");
            } else{
                sb.append(node.val);
                sb.append(",");
                if(node.left == null) q.add(new TreeNode(1001));
                else q.add(node.left);
                if(node.right == null) q.add(new TreeNode(1001));
                else q.add(node.right);
            }
        }
        sb.setLength(sb.length() -1);
        System.out.println(sb.toString());
        return sb.toString();
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        if(data.equals("")) return null;
        String[] arr = data.split(",");
        TreeNode root = new TreeNode(Integer.parseInt(arr[0]));
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        int p = 1;
        while(!q.isEmpty()){
            TreeNode node = q.remove();
            if(!arr[p].equals("#")){
                node.left = new TreeNode(Integer.parseInt(arr[p]));
                p++;
                q.add(node.left);
            }else{
                p++;
            }

            if(!arr[p].equals("#")){
                node.right = new TreeNode(Integer.parseInt(arr[p]));
                p++;
                q.add(node.right);
            }else{
                p++;
            }
        }
        return root;
    }
}

// Your Codec object will be instantiated and called as such:
// Codec ser = new Codec();
// Codec deser = new Codec();
// TreeNode ans = deser.deserialize(ser.serialize(root));